Minor Update to Fix Voyager Bundle Adjustment

dcookastro opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
Added ability to turn off back-of-planet test during bundle adjustment and fixed minor bugs

## Description
Added an option to not do the back-of-planet test so that it can be turned off during bundle adjustment.
  Added optional bool argument to CameraGroundMap::GetXY and RadarGroundMap::GetXY to control 
    whether the test is done or not
  Change the variable name of the angle variable in CameraGroundMap::GetXY to cosAngle to be more descriptive
  Modified call to GetXY in BundleAdjust::ComputePartials, ControlPoint::ComputeResiduals, and ControlPoint::ComputeResidualsMillimeters to not do back-of-planet test
  Changed test value for valid distance/angle from DBL_EPSILON to 1.0e-50 in SurfacePoint methods
    MetersToLongitude, GetLatSigmaDistance, GetLocalRadius, and GetLonSigmaDistance.

## Related Issue
<!--- This project only accepts pull requests related to open issues (GitIssues or RedMine Issues at https://fixit.wr.usgs.gov/fixit) -->
<!--- If suggesting a new feature or change, please discuss it in an issue first -->
<!--- If fixing a bug, there should be an issue describing it with steps to reproduce -->
<!--- Please link to the issue here: -->https://github.com/USGS-Astrogeology/ISIS3/issues/2591

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->These changes fix known problems with running jigsaw on control nets imported from Isis2.

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->Known cases that failed ran successfully after the changes.
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->All system tests were run and passed excluding existing problems with the current nightly build.

## Screenshots (if appropriate):

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ x] Bug fix (non-breaking change which fixes an issue)
- [ x] New feature (non-breaking change which adds functionality)
- [ ] Documentation change (update to the documentation; no code change)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
<!--- - [ x] My code follows the code style of this project. -->
- [ x] I have read and agree to abide by the [Code of Conduct](https://usgs-astrogeology.github.io/code/)
- [ x] I have read the [**CONTRIBUTING**](https://github.com/USGS-Astrogeology/ISIS3/blob/dev/CONTRIBUTING.md) document.
- [ x] My change requires a change to the documentation.
- [ x] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ x] All new and existing tests passed.

## Licensing
This project is mostly composed of free and unencumbered software released into the public domain, and we are unlikely to accept contributions that are not also released into the public domain. Somewhere near the top of each file should have these words:

> This work is free and unencumbered software released into the public domain. In jurisdictions that recognize copyright laws, the author or authors of this software dedicate any and all copyright interest in the software to the public domain.

- [x ] I dedicate any and all copyright interest in this software to the public domain. I make this dedication for the benefit of the public at large and to the detriment of my heirs and successors. I intend this dedication to be an overt act of relinquishment in perpetuity of all present and future rights to this software under copyright law.

dcookastro committed:
Added argument to CameraGroundMap::GetXY to allow back-of-planet test to be skipped for the bundle adjust

dcookastro committed:
Merge branch 'dev' of github.com:USGS-Astrogeology/ISIS3 into dev

dcookastro committed:
Added ability to turn off back-of-planet test while iterating in a bundle adjust

dcookastro committed:
Merge branch 'dev' of github.com:USGS-Astrogeology/ISIS3 into dev

dcookastro committed:
Merge branch 'dev' of github.com:USGS-Astrogeology/ISIS3 into dev

dcookastro committed:
Replaced DBL_EPSILON as a testing constant for valid distances and angles in SurfacePoint.

dcookastro committed:
Merge branch 'dev' of github.com:USGS-Astrogeology/ISIS3 into dev
Merging dev into my clone to get current in preparation for a PR.

dcookastro committed:
Merge branch 'dev' of github.com:USGS-Astrogeology/ISIS3 into dev

jessemapel commented on isis/src/base/objs/CameraGroundMap/CameraGroundMap.cpp:
```diff
@@ -193,28 +195,42 @@ namespace Isis {
-    else {
-      emission = acos(angle) * 180.0 / Isis::PI;
-    }
+    if (test == true) {
```
This should just be if (test)

jessemapel commented on isis/src/base/objs/CameraGroundMap/CameraGroundMap.cpp:
```diff
@@ -254,6 +270,7 @@ namespace Isis {
     SurfacePoint spoint(Latitude(lat, Angle::Degrees),
                         Longitude(lon, Angle::Degrees),
                         Distance(radius, Distance::Meters));
+    // This is called by the application socetlinescankeywords
```
This should be part of the function documentation, maybe an @see tag?

jessemapel commented on isis/src/base/objs/RadarGroundMap/RadarGroundMap.h:
```diff
@@ -117,7 +120,7 @@ namespace Isis {
       virtual bool SetGround(const SurfacePoint &surfacePoint);
       virtual bool GetXY(const SurfacePoint &spoint, double *cudx,
-                         double *cudy);
+                         double *cudy, bool test=false);
```
This is going to be accessed polymorphically through the parent CameraGroundMap class which means `test` will default to true instead of false.

https://stackoverflow.com/questions/5273144/c-polymorphism-and-default-argument

I'm not sure what the correct solution is here.

jessemapel reviewed (commented):
Browsing the code base, it looks like the places where GetXY gets called are:

BundleAdjust
ControlPoint
socetlinescankeywords

The new argument is set to false everywhere but socetlinescankeywords. Should it be set false there? If so, does it make sense to just remove the check completely because it won't ever get called with true?

dcookastro commented:
Jesse,

Thanks for reviewing my changes.  The reason I added the argument was to
keep the test for socetlinescankeywords.
I fixed the (test == true) statement and added the @see statement. I
realize the default will come from the parent, but I wanted to indicate
that it should be set to false.  I forced it to be false in all the
existing Isis calls.  Would it be better just to put a comment instead?

On Thu, Jun 27, 2019 at 12:15 PM Jesse Mapel <notifications@github.com>
wrote:

> *@jessemapel* commented on this pull request.
>
> Browsing the code base, it looks like the places where GetXY gets called
> are:
>
> BundleAdjust
> ControlPoint
> socetlinescankeywords
>
> The new argument is set to false everywhere but socetlinescankeywords.
> Should it be set false there? If so, does it make sense to just remove the
> check completely because it won't ever get called.
> ------------------------------
>
> In isis/src/base/objs/CameraGroundMap/CameraGroundMap.cpp
> <https://github.com/USGS-Astrogeology/ISIS3/pull/3343#discussion_r298325667>
> :
>
> > -    double upsB[3], upB[3], dist;
> -    vminus_c((SpiceDouble *) &lookB[0], upsB);
> -    unorm_c(upsB, upsB, &dist);
> -    unorm_c((SpiceDouble *) &pB[0], upB, &dist);
> -    double angle = vdot_c(upB, upsB);
> -    double emission;
> -    if (angle > 1) {
> -      emission = 0;
> -    }
> -    else if (angle < -1) {
> -      emission = 180.;
> -    }
> -    else {
> -      emission = acos(angle) * 180.0 / Isis::PI;
> -    }
> +    if (test == true) {
>
> This should just be if (test)
> ------------------------------
>
> In isis/src/base/objs/CameraGroundMap/CameraGroundMap.cpp
> <https://github.com/USGS-Astrogeology/ISIS3/pull/3343#discussion_r298325896>
> :
>
> > @@ -254,6 +270,7 @@ namespace Isis {
>      SurfacePoint spoint(Latitude(lat, Angle::Degrees),
>                          Longitude(lon, Angle::Degrees),
>                          Distance(radius, Distance::Meters));
> +    // This is called by the application socetlinescankeywords
>
> This should be part of the function documentation, maybe an @see
> <https://github.com/see> tag?
> ------------------------------
>
> In isis/src/base/objs/RadarGroundMap/RadarGroundMap.h
> <https://github.com/USGS-Astrogeology/ISIS3/pull/3343#discussion_r298327290>
> :
>
> > @@ -117,7 +120,7 @@ namespace Isis {
>        virtual bool SetGround(const Latitude &lat, const Longitude &lon);
>        virtual bool SetGround(const SurfacePoint &surfacePoint);
>        virtual bool GetXY(const SurfacePoint &spoint, double *cudx,
> -                         double *cudy);
> +                         double *cudy, bool test=false);
>
> This is going to be accessed polymorphically through the parent
> CameraGroundMap class which means test will default to true instead of
> false.
>
>
> https://stackoverflow.com/questions/5273144/c-polymorphism-and-default-argument
>
> I'm not sure what the correct solution is here.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/USGS-Astrogeology/ISIS3/pull/3343?email_source=notifications&email_token=AIEV3S4TIPMLASAC7KTWZTDP4UGS3A5CNFSM4H3643CKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOB44W7KA#pullrequestreview-255422376>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AIEV3S5G4QUM2IPOZBFWZ5LP4UGS3ANCNFSM4H3643CA>
> .
>

dcookastro committed:
Corrected a bool check in CameraGroundMap::GetXY

jessemapel commented:
What is the current functionality in the radar ground map? It doesn't attempt to do any type of checking? I would be tempted to just ignore the argument in the radar ground map and document as such. That doesn't seem like a good solution either though.

Right now, anything that calls camera.groundmap().GetXY() for a radar sensor will succeed, but with this change, that code will hit the new exception.

dcookastro commented:
My motivation in adding the exception was to alert the programmer that the
test was not being applied despite the bool being set to true. The current
functionality of radar ground map is to not do the back-of-planet test or
care whether it was done or not. I guess I wanted to make sure anyone using
the method thought about what they were doing, since the behavior differs
from the parent class. I agree this is not a good solution.  Ken and I
discussed this before he left too, and I could not come up with a good
solution.

On Thu, Jun 27, 2019 at 1:35 PM Jesse Mapel <notifications@github.com>
wrote:

> What is the current functionality in the radar ground map? It doesn't
> attempt to do any type of checking? I would be tempted to just ignore the
> argument in the radar ground map and document as such. That doesn't seem
> like a good solution either though.
>
> Right now, anything that calls camera.groundmap().GetXY() for a radar
> sensor will succeed, but with this change, that code will hit the new
> exception.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/USGS-Astrogeology/ISIS3/pull/3343?email_source=notifications&email_token=AIEV3S244KNFEJHGS7XXJEDP4UP6TA5CNFSM4H3643CKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODYYJN4I#issuecomment-506500849>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AIEV3S7VH6I6TKNH7BHAQOTP4UP6TANCNFSM4H3643CA>
> .
>

jessemapel commented:
I'm okay with going forward with this then.

scsides reviewed (approved):
Agree with the conversation about testing the backside. Can't say I can think of a better way either. Everything else looks good.

scsides commented:
sgstapleton not available, so I'll go ahead and merge
